macOS static build support (#4163)
mixxxdj/mixxx: CMakeLists.txt

  target_include_directories(mixxx PRIVATE "${CMAKE_CURRENT_SOURCE_DIR}")
endif()

# Chromaprint
find_package(Chromaprint REQUIRED)
target_link_libraries(mixxx-lib PRIVATE Chromaprint::Chromaprint)

# Locale Aware Compare for SQLite
find_package(SQLite3)

> # For LOCALECOMPAREwe call directly sqlite functions to the database opened by

Reviewer: ```suggestion
# For LOCALECOMPARE we call directly sqlite functions to the database opened by
```

# Qt. It only works without crashing when Mixxx links to the same sqlite
# library as Qt.

> # This is difficult on macOS where system SQLite, can be installed and linked

Reviewer: ```suggestion
# This is difficult on macOS where system the SQLite can be installed and linked
```

# dynamically from various locations. There is no issue in case of static
# linking which would result in a duplicate symbol error.
if(NOT SQLite3_FOUND)
  set(LOCALECOMPARE_DEFAULT OFF)
else()
  is_static_library(SQLite3_IS_STATIC SQLite3::SQLite3)
  if(SQLite3_IS_STATIC OR NOT APPLE)
    set(LOCALECOMPARE_DEFAULT ON)
  else()
    set(LOCALECOMPARE_DEFAULT OFF)
  endif()
endif()
cmake_dependent_option(LOCALECOMPARE "Locale Aware Compare support for SQLite" ON "LOCALECOMPARE_DEFAULT" OFF)
if(LOCALECOMPARE)
  if(NOT SQLite3_FOUND)
    message(FATAL_ERROR "Locale Aware Compare for SQLite requires libsqlite and its development headers.")
  endif()
  target_compile_definitions(mixxx-lib PUBLIC __SQLITE3__)
  target_link_libraries(mixxx-lib PRIVATE SQLite3::SQLite3)
elseif(SQLite3_IS_STATIC)

>   # in the static case we need to link SQlite3 uncoditionally

Reviewer: ```suggestion
  # in the static case we need to link SQLite3 uncoditionally
```

Reviewer 2: and typo: unconditionally

find_package(Chromaprint REQUIRED)
target_link_libraries(mixxx-lib PRIVATE Chromaprint::Chromaprint)

# Locale Aware Compare for SQLite
find_package(SQLite3)
# For LOCALECOMPAREwe call directly sqlite functions to the database opened by
# Qt. It only works without crashing when Mixxx links to the same sqlite
# library as Qt.
# This is difficult on macOS where system SQLite, can be installed and linked
# dynamically from various locations. There is no issue in case of static
# linking which would result in a duplicate symbol error.
if(NOT SQLite3_FOUND)
  set(LOCALECOMPARE_DEFAULT OFF)
else()
  get_target_property(SQLITE3_TYPE SQLite3::SQLite3 TYPE)
  if(SQLITE3_TYPE STREQUAL "STATIC_LIBRARY" OR NOT APPLE)
    set(LOCALECOMPARE_DEFAULT ON)
  else()
    set(LOCALECOMPARE_DEFAULT OFF)
  endif()
endif()
cmake_dependent_option(LOCALECOMPARE "Locale Aware Compare support for SQLite" ON "LOCALECOMPARE_DEFAULT" OFF)
if(LOCALECOMPARE)
  if(NOT SQLite3_FOUND)
    message(FATAL_ERROR "Locale Aware Compare for SQLite requires libsqlite and its development headers.")
  endif()
  target_compile_definitions(mixxx-lib PUBLIC __SQLITE3__)
  target_link_libraries(mixxx-lib PRIVATE SQLite3::SQLite3)

> elseif(CHROMAPRINT_TYPE STREQUAL "STATIC_LIBRARY")

Reviewer: What does Chromaprint have to do with SQLite?

target_link_libraries(mixxx-lib PRIVATE ${PortMidi_LIBRARIES})

# Protobuf

> if(NOT BUILD_SHARED_LIBS)

Reviewer 3: I suppose this caused the issue reported on here? https://mixxx.zulipchat.com/#narrow/stream/247620-development-help/topic/cmake.20fails.20with.20Could.20NOT.20find.20Protobuf

Reviewer: Fixed in #4185

  Qt5::Widgets
  Qt5::Xml)
target_compile_definitions(mixxx-lib PUBLIC QT_TABLET_SUPPORT QT_USE_QSTRINGBUILDER)

> is_static_library(Qt5_IS_STATIC Qt5::Core)

Reviewer: ```suggestion
is_static_library(Qt_IS_STATIC Qt5::Core)
```
to avoid needing to rename this when adding Qt6 support

Author: Qt5 reflects the target name as everywhere else. I like to keep it consistent.

Reviewer: okay, I'll rename this when rebasing #4052

if(APPLE)
  if(QT5_TYPE STREQUAL "STATIC_LIBRARY")
    find_package(FFTW REQUIRED)

>     find_package(SndFile REQUIRED)

Reviewer: already linked below

>     find_package(sord CONFIG REQUIRED)

Reviewer: handled by Findlilv.cmake

    find_package(SQLite3 REQUIRED)
    find_library(SAMPLERATE_LIBRARY samplerate REQUIRED)
    target_link_libraries(mixxx-lib PRIVATE
        FFTW::FFTW
        SndFile::sndfile
        sord::sord
        ${SQLite3_LIBRARIES}
        ${SAMPLERATE_LIBRARY}

>     )

Reviewer: Aren't these handled elsewhere in CMakeLists.txt?

Author: I have tried to get rid of some of them but these are the remaining bits.

endif()

if(APPLE)
  if(QT5_TYPE STREQUAL "STATIC_LIBRARY")

>     find_package(FFTW REQUIRED)

Reviewer: What is this needed for?

Author: Chromaprint and Rubberband, unfortunately setting INTERFACE_LINK_LIBRARIES is not sufficient.

    find_package(SndFile REQUIRED)
    find_package(sord CONFIG REQUIRED)

>     find_library(SAMPLERATE_LIBRARY samplerate REQUIRED)

Reviewer: I'm confused... Mixxx doesn't use libsamplerate?

Author: rubberband uses libsamplerate

Reviewer 2: ...and Rubberband also needs FFTW, depending on the build config. Rubberband 1.9.2 comes with its own, optimized FFT code that doesn't require external dependencies.

endif()

if(APPLE)
  if(Qt5_IS_STATIC)
    target_link_libraries(mixxx-lib PRIVATE
        "-weak_framework Accelerate"
        "-weak_framework AppKit"
        "-weak_framework AudioToolbox"
        "-weak_framework AudioUnit"
        "-weak_framework AVFoundation"
        "-weak_framework CoreAudio"
        "-weak_framework CoreFoundation"
        "-weak_framework CoreImage"
        "-weak_framework CoreMedia"
        "-weak_framework CoreMidi"
        "-weak_framework CoreServices"
        "-weak_framework CoreVideo"
        "-weak_framework IOSurface"
        "-weak_framework VideoToolbox"
    )
  else()

>     # Used for battery measurements and controlling the screensaver on OS X and iOS.

Reviewer: ```suggestion
    # Used for battery measurements and controlling the screensaver on macOS.
```
Mixxx does not support iOS.

# FDK-AAC is loaded dynamically at runtime by EncoderFdkAac using QLibrary,
# so copy it into the Windows and macOS packages, but do not link to it.
if(APPLE AND MACOS_BUNDLE)
  find_library(FDK_AAC_LIBRARY fdk-aac)
  if(FDK_AAC_LIBRARY)
    message(STATUS "Found fdk-aac: ${FDK_AAC_LIBRARY}")

>     install(FILES ${FDK_AAC_LIBRARY} DESTINATION ${MIXXX_INSTALL_PREFIX}/Contents/Frameworks FOLLOW_SYMLINK_CHAIN)

Reviewer: I remember using this then reverting it when we initially added fdk-aac but I do not remember why I reverted it.
